File: kotlin/src/main/kotlin/solutions/Day12.kt
package solutions

class Day12: Day {

    override fun part1Solution(input: List<String>): String {
        var answer = 0L
        input.forEach { line ->
            val lineParts = line.split(" ")
            val springs = lineParts[0]
            val groupings = lineParts[1].split(",").map { it.toInt() }
            answer += solve(springs, groupings)
        }
        return answer.toString()
    }

    override fun part2Solution(input: List<String>): String {
        var answer = 0L
        input.forEach { line ->
            val lineParts = line.split(" ")
            val springs = lineParts[0]
            val groupings = lineParts[1].split(",").map { it.toInt() }

            var springAlt = springs
            val groupingsAlt = groupings.toMutableList()

            repeat(4){
                springAlt = "$springAlt?$springs"
                groupingsAlt.addAll(groupings)
            }

            answer += solve(springAlt, groupingsAlt)
        }
        return answer.toString()
    }
    private fun solve(input: String, groups: List<Int>): Long {
        val cache = mutableMapOf<Pair<String, List<Int>>, Long>()
        fun goDeeper(line: String, values: List<Int>, index: Int): Long {
            if (index >= line.length) {
                if (values.isEmpty()) {
                    return 1
                }
                return 0
            }
            var answer = 0L

            // Short circuit
            val substring = line.substring(index..<line.length)
            if (substring to values in cache){
                return cache[substring to values]!!
            }

            if (values.isNotEmpty()) {
                if ((index+(values[0] - 1) < line.length) &&
                    line.substring(index..<(index+values[0])).all{ it == '?' || it == '#' } &&
                    (index+values[0] == line.length || line[index+values[0]] != '#')
                ){
                    answer += goDeeper(line, values.subList(1, values.size), index + values[0] + 1)
                }
                if (line[index] == '#') {
                    cache[substring to values] = answer
                    return answer
                }
                answer += goDeeper(line, values, index + 1)
            } else if (line[index] == '#') {
                cache[substring to values] = 0
                return 0
            } else {
                 answer += goDeeper(line, values, index + 1)
            }
            cache[substring to values] = answer
            return answer
        }
        return goDeeper(input, groups, 0)
    }

}

File: swift/Solutions/Day12.swift
//
//  Day12.swift
//  advent_of_code_2023
//
//  Created by Kushantha Attanayake on 12/11/23.
//

import Foundation

struct Day12: Day {
    func part1Solution(input: [String]) -> String {
        var answer = 0
        var cache = [Checkpoint: Int]()
        
        input.forEach { line in
            let lineParts = line.split(separator: " ")
            let springs = String(lineParts[0])
            let groupings = lineParts[1].split(separator: ",").map { Int($0)! }
            answer += goDeeper(springs, groupings, 0, &cache)
        }
        
        return "\(answer)"
    }
    
    func part2Solution(input: [String]) -> String {
        var answer = 0
        var cache = [Checkpoint: Int]()
        var lineNum = 0
        input.forEach { line in
            let lineParts = line.split(separator: " ")
            let springs = String(lineParts[0])
            var springsAlt = springs
            let groupings = lineParts[1].split(separator: ",").map { Int($0)! }
            var groupingsAlt = groupings
            
            for _ in 1..<5 {
                springsAlt += ("?" + springs)
                groupingsAlt += groupings
            }
            answer += goDeeper(springsAlt, groupingsAlt, 0, &cache)
            lineNum += 1
        }
        
        return "\(answer)"
    }
    
    func goDeeper(_ line: String, _ values: [Int], _ index: Int, _ cache: inout [Checkpoint: Int]) -> Int {
        if index >= line.count {
            if values.isEmpty {
                return 1
            }
            return 0
        }
        
        var answer = 0
        let checkpoint = Checkpoint(substring: String(line[index...]), groupings: values)
        if let cachedValue = cache[checkpoint] {
            return cachedValue
        }
        
        if !values.isEmpty {
            if (index+(values[0] - 1) < line.count),
               line[index..<(index+values[0])].allSatisfy ({ $0 == "?" || $0 == "#" }),
               (index+values[0] == line.count || line[index+values[0]] != "#")
            {
                answer += goDeeper(line, values[1...].map { $0 }, index + values[0] + 1, &cache)
            }
            if line[index] == "#" {
                cache[checkpoint] = answer
                return answer
            }
            answer += goDeeper(line, values, index + 1, &cache)
        } else if line[index] == "#" {
            cache[checkpoint] = answer
            return answer
        } else {
            answer += goDeeper(line, values, index + 1, &cache)
        }
        cache[checkpoint] = answer
        return answer
    }
    
}

struct Checkpoint: Hashable {
    let substring: String
    let groupings: [Int]
}
